x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/persistence/DataFrameTransformsCheckpointService.java

```diff
@@ -0,0 +1,97 @@
+ *
+ * This will be used to checkpoint a transform, detect changes, run the transform in continuous mode.
+ *
+ */
+public class DataFrameTransformsCheckpointService {
+
+    private static final Logger logger = LogManager.getLogger(DataFrameTransformsCheckpointService.class);
+
+    private final Client client;
+
+    public DataFrameTransformsCheckpointService(final Client client) {
+        this.client = client;
+    }
+
+    /**
+     * Get a checkpoint that is not attached to an id. E.g. for change detection.
+     *
+     * @param transformConfig the @link{DataFrameTransformConfig}
+     * @param listener listener to call after inner request returned
+     */
+    public void getCheckpoint(DataFrameTransformConfig transformConfig, ActionListener<DataFrameTransformCheckpoint> listener) {
+        getCheckpoint(transformConfig, -1L, listener);
+    }
+
+    /**
+     * Get a checkpoint with an id, used to store a checkpoint.
+     *
+     * @param transformConfig the @link{DataFrameTransformConfig}
+     * @param listener listener to call after inner request returned
+     */
+    public void getCheckpoint(DataFrameTransformConfig transformConfig, long checkpointId,
+            ActionListener<DataFrameTransformCheckpoint> listener) {
+        long timestamp = System.currentTimeMillis();
+
+        // placeholder for time based synchronization
+        long timeUpperBound = 0;
+
+        ClientHelper.executeWithHeadersAsync(transformConfig.getHeaders(), ClientHelper.DATA_FRAME_ORIGIN, client,
+                IndicesStatsAction.INSTANCE, new IndicesStatsRequest().indices(transformConfig.getSource()),
+                ActionListener.wrap(response -> {
+                    Map<String, long[]> checkpointsByIndex = extractIndexCheckPoints(response.getIndices());
+                    DataFrameTransformCheckpoint checkpointDoc = new DataFrameTransformCheckpoint(transformConfig.getId(), timestamp,
+                            checkpointId, checkpointsByIndex, timeUpperBound);
+                    listener.onResponse(checkpointDoc);
+
+                }, IndicesStatsRequestException -> {
+                    throw new RuntimeException("Failed to retrieve indices stats", IndicesStatsRequestException);
+                }));
+    }
+
+    private static Map<String, long[]> extractIndexCheckPoints(Map<String, IndexStats> indexStatsByIndex) {
+        Map<String, long[]> checkpointsByIndex = new TreeMap<>();
+        for (Entry<String, IndexStats> stats : indexStatsByIndex.entrySet()) {
+            String indexName = stats.getKey();
+            List<Long> checkpoints = new ArrayList<>();
+            for (IndexShardStats indexShardStats : stats.getValue()) {
+                for (ShardStats shardStats : indexShardStats.getShards()) {
+                    // we take the global checkpoint, which is consistent across all replicas
+                    checkpoints.add(shardStats.getSeqNoStats().getGlobalCheckpoint());
```

Comment:
Can the order of the elements in `indexShardStats.getShards()` change? Would that then look like a different check point

Reply:
I think this is ordered by shard, but after another look I am unsure about it, will test again.

Reply:
I checked, this is already below the shard level, but this array contains both primary as well as replica. 

Therefore thanks for pointing me to it, this isn't correct as it will not work if the source index has replicas.

Reply:
I addressed this, extraction is now independent of order and it works independent of the number of primary/replica's